infrastructure/lib/constructs/authentication.ts

```diff
@@ -0,0 +1,249 @@
+   * provided.
+   */
+  public readonly idps: cognito.CfnUserPoolIdentityProvider[] = [];
+
+  constructor(scope: cdk.Construct, id: string, props: CognitoAuthenticationProps) {
+    super(scope, id);
+    // The Pre-Token Generation Function is responsible for setting attributes on the User resource
+    // prior to the token being generated. This is used here for setting the groups based on the
+    // groups specified in the SAML Response or OIDC Claim.
+    this.preTokenGenerationFunction = new lambdaNodeJs.NodejsFunction(this, "PreTokenGeneration", {
+      entry: packageRoot() + "/cognito/preTokenGeneration/index.ts",
+      environment: {
+        GROUPS_ATTRIBUTE_CLAIM_NAME: `custom:${props.groupsAttributeName ?? "groups"}`,
+      },
+    });
+
+    this.userPool = new cognito.UserPool(this, "Pool", {
+      signInAliases: {},
+      lambdaTriggers: { preTokenGeneration: this.preTokenGenerationFunction },
+    });
+    const cfnPool = this.userPool.node.defaultChild as CfnUserPool;
+    cfnPool.schema = [
+      {
+        name: props.groupsAttributeName ?? "groups",
+        attributeDataType: "String",
+        mutable: true,
+        required: false,
+        stringAttributeConstraints: {
+          maxLength: "2000",
+        },
+      },
+    ];
+
+    this.adminsGroup = new cognito.CfnUserPoolGroup(this, "AdminsGroup", {
+      groupName: props.adminsGroupName,
+      userPoolId: this.userPool.userPoolId,
+    });
+    this.usersGroup = new cognito.CfnUserPoolGroup(this, "UsersGroup", {
+      groupName: props.usersGroupName,
+      userPoolId: this.userPool.userPoolId,
+    });
+
+    // This allows us to have any number of supported backing IdPs leveraging
+    // OIDC or SAML as the use case requires. The most common scenario is likely
+    // to be a single OIDC IdP though.
+    this.idps.push(...this.processSamlIdps(props.samlIdps ?? [], props));
+    this.idps.push(...this.processOidcIdps(props.oidcIdps ?? [], props));
+
+    this.userPoolDomain = new cognito.UserPoolDomain(this, "CognitoDomain", {
+      cognitoDomain: {
+        domainPrefix: props.cognitoDomain,
+      },
+      userPool: this.userPool,
+    });
+  }
+
+  private processSamlIdps(
+    idps: SAMLIdentityProviderProps[],
+    props: CognitoAuthenticationProps
+  ): cognito.CfnUserPoolIdentityProvider[] {
+    return idps.map(
+      (idp) =>
+        new cognito.CfnUserPoolIdentityProvider(this, "SamlIdp" + idp.providerName, {
+          providerName: idp.providerName,
+          providerDetails: {
+            MetadataURL: idp.metadataURL,
+          },
+          attributeMapping: { ...samlAttributeMapping(props) },
+          providerType: AuthenticationProtocol.SAML,
+          userPoolId: this.userPool.userPoolId,
+        })
+    );
+  }
+
+  private processOidcIdps(
+    idps: OIDCIdentityProviderProps[],
+    props: CognitoAuthenticationProps
+  ): cognito.CfnUserPoolIdentityProvider[] {
+    return idps.map(
+      (idp) =>
```

Comment:
I have done this exact sort of thing in Java (creating the generic interfaces to capture similar configs then mapping them to some library object or whatever) so you won't see any eye-rolling about this being overkill from me. :)